conan/tools/cmake/toolchain/blocks.py

```diff
@@ -574,20 +555,95 @@ class TryCompileBlock(Block):
     template = textwrap.dedent("""
         {% if cmake_system_name %}
         # Cross building
         set(CMAKE_SYSTEM_NAME {{ cmake_system_name }})
         {% endif %}
         {% if cmake_system_version %}
-        # Cross building
         set(CMAKE_SYSTEM_VERSION {{ cmake_system_version }})
         {% endif %}
         {% if cmake_system_processor %}
         set(CMAKE_SYSTEM_PROCESSOR {{ cmake_system_processor }})
         {% endif %}
+        """)
+
+    def _get_generic_system_name(self):
+        os_host = self._conanfile.settings.get_safe("os")
+        os_build = self._conanfile.settings_build.get_safe("os")
+        arch_host = self._conanfile.settings.get_safe("arch")
+        arch_build = self._conanfile.settings_build.get_safe("arch")
+        cmake_system_name_map = {"Neutrino": "QNX",
+                                 "": "Generic",
+                                 None: "Generic"}
+        if os_host != os_build:
+            return cmake_system_name_map.get(os_host, os_host)
+        elif arch_host is not None and arch_host != arch_build:
+            if not ((arch_build == "x86_64") and (arch_host == "x86") or
+                    (arch_build == "sparcv9") and (arch_host == "sparc") or
+                    (arch_build == "ppc64") and (arch_host == "ppc32")):
+                return cmake_system_name_map.get(os_host, os_host)
+
+    def _is_apple_cross_building(self):
+        os_host = self._conanfile.settings.get_safe("os")
+        arch_host = self._conanfile.settings.get_safe("arch")
+        arch_build = self._conanfile.settings_build.get_safe("arch")
+        return os_host in ('iOS', 'watchOS', 'tvOS') or (os_host == 'Macos' and arch_host != arch_build)
+
+    def _system_name(self):
+        os_host = self._conanfile.settings.get_safe("os")
+        if self._is_apple_cross_building():
+            # cross-build in Macos also for M1
+            return {'Macos': 'Darwin'}.get(os_host, os_host)
+        elif os_host != 'Android':
+            return self._get_generic_system_name()
+
+    def _system_version(self):
+        os_host = self._conanfile.settings.get_safe("os")
+        if self._is_apple_cross_building():
+            #  CMAKE_SYSTEM_VERSION for Apple sets the sdk version, not the os version
+            return self._conanfile.settings.get_safe("os.sdk_version")
+        elif os_host != 'Android':
+            return self._conanfile.settings.get_safe("os.version")
+
+    def _system_processor(self):
+        os_host = self._conanfile.settings.get_safe("os")
+        arch_host = self._conanfile.settings.get_safe("arch")
+        arch_build = self._conanfile.settings_build.get_safe("arch")
+        if self._is_apple_cross_building():
+            return to_apple_arch(arch_host)
+        elif os_host != 'Android' and arch_host != arch_build:
+            return arch_host
+
+    def _get_cross_build(self):
+        user_toolchain = self._conanfile.conf.get("tools.cmake.cmaketoolchain:user_toolchain")
```

Comment:
Maybe (not related to this PR, just a possible improvement) it is worth honoring the other confs, and letting users passing both the ``user_toolchain`` and the other ``system_xxxx```confs. This can be done by swapping place and doing a ``return system_name, system_version, system_processor``.


The thing is that it would be weird to define ``tools.cmake.cmaketoolchain:system_name`` and that being completely rejected if you have ``tools.cmake.cmaketoolchain:user_toolchain`` defined?